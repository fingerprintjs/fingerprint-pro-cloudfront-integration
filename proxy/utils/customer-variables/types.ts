export enum CustomerVariableType {
  BehaviourPath = 'fpjs_behavior_path',
  GetResultPath = 'fpjs_get_result_path',
  PreSharedSecret = 'fpjs_pre_shared_secret',
  AgentDownloadPath = 'fpjs_agent_download_path',
}

export type CustomerVariableValue = string | null | undefined

export type CustomerVariablesRecord = Record<CustomerVariableType, CustomerVariableValue>

export interface CustomerVariableProvider {
  readonly name: string

  getVariable: (variable: CustomerVariableType) => Promise<CustomerVariableValue>
}
