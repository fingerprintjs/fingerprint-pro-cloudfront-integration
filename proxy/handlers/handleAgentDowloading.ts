import { AgentOptions } from '../model'
import { CloudFrontResultResponse } from 'aws-lambda'
import https from 'https'

import { updateResponseHeadersForAgentDownload, addTrafficMonitoringSearchParamsForProCDN } from '../utils'

export function downloadAgent(options: AgentOptions): Promise<CloudFrontResultResponse> {
  return new Promise((resolve) => {
    const data: any[] = []

    const url = new URL('https://__FPCDN__')
    url.pathname = getEndpoint(options.apiKey, options.version, options.loaderVersion)
    addTrafficMonitoringSearchParamsForProCDN(url)

    options.logger.debug(`Downloading agent from: ${url.toString()}`)

    const request = https.request(
      url,
      {
        method: options.method,
        headers: options.headers,
      },
      (response) => {
        let binary = false
        if (response.headers['content-encoding']) {
          binary = true
        }

        response.setEncoding(binary ? 'binary' : 'utf8')

        response.on('data', (chunk) => data.push(Buffer.from(chunk, 'binary')))

        response.on('end', () => {
          const body = Buffer.concat(data)
          resolve({
            status: response.statusCode ? response.statusCode.toString() : '500',
            statusDescription: response.statusMessage,
            headers: updateResponseHeadersForAgentDownload(response.headers, options.domain),
            bodyEncoding: 'base64',
            body: body.toString('base64'),
          })
        })
      },
    )

    request.on('error', (error) => {
      options.logger.error('unable to download agent', { error })
      resolve({
        status: '500',
        statusDescription: 'Bad request',
        headers: {},
        bodyEncoding: 'text',
        body: 'error',
      })
    })

    request.end()
  })
}

function getEndpoint(apiKey: string | undefined, version: string, loaderVersion: string | undefined): string {
  const lv: string = loaderVersion !== undefined && loaderVersion !== '' ? `/loader_v${loaderVersion}.js` : ''
  return `/v${version}/${apiKey}${lv}`
}
