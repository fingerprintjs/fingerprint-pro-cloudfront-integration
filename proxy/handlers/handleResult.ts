import { ResultOptions } from '../model'
import { CloudFrontResultResponse } from 'aws-lambda'
import https from 'https'
import { Region } from '../model'

import { updateResponseHeaders, addTrafficMonitoringSearchParamsForVisitorIdRequest } from '../utils'

export function handleResult(options: ResultOptions): Promise<CloudFrontResultResponse> {
  return new Promise((resolve) => {
    console.debug('Handling result:', { options })

    const data: any[] = []

    const url = new URL(getIngressAPIHost(options.region, options.fpIngressBaseHost) + options.suffix)
    decodeURIComponent(options.querystring)
      .split('&')
      .filter((it) => it.includes('='))
      .forEach((it) => {
        const kv = it.split('=')
        if (kv[0] === 'region') {
          kv[1] = options.region
        }
        url.searchParams.append(kv[0], kv[1])
      })
    if (options.method === 'POST') {
      addTrafficMonitoringSearchParamsForVisitorIdRequest(url)
    }

    console.debug(`Performing request: ${url.toString()}`)

    const request = https.request(
      url,
      {
        method: options.method,
        headers: options.headers,
      },
      (response) => {
        response.on('data', (chunk) => data.push(chunk))

        response.on('end', () => {
          const payload = Buffer.concat(data)

          console.debug('Response from Ingress API', {
            statusCode: response.statusCode,
            payload: payload.toString('utf-8'),
          })

          resolve({
            status: response.statusCode ? response.statusCode.toString() : '500',
            statusDescription: response.statusMessage,
            headers: updateResponseHeaders(response.headers),
            bodyEncoding: 'base64',
            body: payload.toString('base64'),
          })
        })
      }
    )

    request.write(Buffer.from(options.body, 'base64'))

    request.on('error', (error) => {
      console.error('unable to handle result', { error })
      resolve({
        status: '500',
        statusDescription: 'Bad request',
        headers: {},
        bodyEncoding: 'text',
        body: generateErrorResponse(error),
      })
    })

    request.end()
  })
}

function generateErrorResponse(err: Error): string {
  const body = {
    v: '2',
    error: {
      code: 'Failed',
      message: `An error occurred with Fingerprint Pro Lambda function. Reason ${err}`,
    },
    requestId: generateRequestId,
    products: {},
  }
  return JSON.stringify(body)
}

function generateRequestId(): string {
  const uniqueId = generateRequestUniqueId()
  const now = new Date().getTime()
  return `${now}.aws-${uniqueId}`
}

function generateRandomString(length: number): string {
  let result = ''
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * characters.length))
  }
  return result
}

function generateRequestUniqueId(): string {
  return generateRandomString(2)
}

function getIngressAPIHost(region: Region, baseHost: string): string {
  const prefix = region === Region.us ? '' : `${region}.`
  return `https://${prefix}${baseHost}`
}
