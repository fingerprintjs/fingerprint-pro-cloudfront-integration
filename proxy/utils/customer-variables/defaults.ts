import { CustomerVariableValue, CustomerVariableType } from './types'

const defaultCustomerVariables: Record<CustomerVariableType, CustomerVariableValue> = {
  [CustomerVariableType.BehaviourPath]: 'fpjs',
  [CustomerVariableType.GetResultPath]: 'resultId',
  [CustomerVariableType.PreSharedSecret]: null,
  [CustomerVariableType.AgentDownloadPath]: 'agent',
}

export function getDefaultCustomerVariable(variable: CustomerVariableType): CustomerVariableValue {
  return defaultCustomerVariables[variable]
}
