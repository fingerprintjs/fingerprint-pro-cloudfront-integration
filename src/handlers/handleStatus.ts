import { CloudFrontResultResponse } from 'aws-lambda'

export function handleStatus(): Promise<CloudFrontResultResponse> {
  return new Promise((resolve) => {
    //todo add content (version, internal settings verification)
    const body = {
      version: '__lambda_func_version__',
    }
    resolve({
      status: '200',
      body: JSON.stringify(body),
    })
  })
}
