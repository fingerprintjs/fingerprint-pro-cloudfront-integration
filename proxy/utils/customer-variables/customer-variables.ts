import { CustomerVariableProvider, CustomerVariableType, CustomerVariableValue } from './types'
import { getDefaultCustomerVariable } from './defaults'
import { createLogger } from '../../logger'

export interface GetVariableResult {
  value: CustomerVariableValue
  resolvedBy: string | null
}

/**
 * Allows access to customer defined variables using multiple providers.
 * Variables will be resolved in order in which providers are set.
 * */
export class CustomerVariables {
  constructor(
    private readonly providers: CustomerVariableProvider[],
    private readonly logger = createLogger(),
  ) {}

  /**
   * Attempts to resolve customer variable using providers.
   * If no provider can resolve the variable, the default value is returned.
   * */
  async getVariable(variable: CustomerVariableType): Promise<GetVariableResult> {
    const providerResult = await this.getValueFromProviders(variable)

    if (providerResult) {
      return providerResult
    }

    const defaultValue = getDefaultCustomerVariable(variable)

    this.logger.debug(`Resolved customer variable ${variable} with default value ${defaultValue}`)

    return {
      value: defaultValue,
      resolvedBy: null,
    }
  }

  private async getValueFromProviders(variable: CustomerVariableType): Promise<GetVariableResult | null> {
    for (const provider of this.providers) {
      try {
        const result = await provider.getVariable(variable)

        if (result) {
          this.logger.debug(`Resolved customer variable ${variable} with provider ${provider.name}`)

          return {
            value: result,
            resolvedBy: provider.name,
          }
        }
      } catch (error) {
        this.logger.error(`Error while resolving customer variable ${variable} with provider ${provider.name}`, {
          error,
        })
      }
    }

    return null
  }
}
