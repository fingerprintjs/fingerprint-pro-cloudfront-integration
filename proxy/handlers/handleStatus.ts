import { CloudFrontResultResponse } from 'aws-lambda'
import { CustomerVariables } from '../utils/customer-variables/customer-variables'
import { CustomerVariableType, CustomerVariableValue } from '../utils/customer-variables/types'
import { maybeObfuscateVariable } from '../utils/customer-variables/maybe-obfuscate-variable'

export interface EnvVarInfo {
  envVarName: string
  value: CustomerVariableValue
  isSet: boolean
  // If null, the variable was resolved with the default value, otherwise it was resolved by the provider with the given name
  resolvedBy: string | null
}

export interface StatusInfo {
  version: string
  envInfo: EnvVarInfo[]
}

async function getEnvInfo(customerVariables: CustomerVariables) {
  const infoArray: EnvVarInfo[] = await Promise.all(
    Object.values(CustomerVariableType).map(async (variable) => {
      const value = await maybeObfuscateVariable(customerVariables, variable)

      return {
        envVarName: variable,
        value: value.value,
        isSet: Boolean(value.value),
        resolvedBy: value.resolvedBy,
      }
    })
  )

  return infoArray
}

function renderEnvInfo(envInfo: EnvVarInfo[]) {
  const isAlSet = envInfo.every((info) => info.isSet && info.resolvedBy)

  if (isAlSet) {
    return `
      <div>
        ✅ All environment variables are set
      </div>
    `
  }

  const children = envInfo
    .filter((info) => !info.isSet || !info.resolvedBy)
    .map(
      (info) => `
        <div class="env-info-item">
            ⚠️ <strong>${info.envVarName} </strong> is not defined${info.isSet ? ' and uses default value' : ''}
        </div>`
    )

  return `
    <div class="env-info">
      ${children.join('')}
    </div>
  `
}

function renderHtml({ version, envInfo }: StatusInfo) {
  return `
    <html lang="en-US">
      <head>
        <title>CloudFront integration status</title>
        <meta charset="utf-8">
        <style>
          body, .env-info {
            display: flex;
          }
          
          body {
            flex-direction: column;
            align-items: center;
          }
          
          body > * {
            margin-bottom: 1em;
          }
        </style>
      </head>
      <body>
        <h1>CloudFront integration status</h1>
        <div>
          Lambda function version: ${version}
        </div>
        ${renderEnvInfo(envInfo)}
          <span>
            Please reach out our support via <a href="mailto:support@fingerprint.com">support@fingerprint.com</a> if you have any issues
          </span>
      </body>
    </html>
  `
}

export async function getStatusInfo(customerVariables: CustomerVariables): Promise<StatusInfo> {
  return {
    version: '__lambda_func_version__',
    envInfo: await getEnvInfo(customerVariables),
  }
}

export async function handleStatus(customerVariables: CustomerVariables): Promise<CloudFrontResultResponse> {
  const body = await getStatusInfo(customerVariables)

  return {
    status: '200',
    body: renderHtml(body).trim(),
    headers: {
      'content-type': [{ key: 'Content-Type', value: 'text/html' }],
    },
  }
}
