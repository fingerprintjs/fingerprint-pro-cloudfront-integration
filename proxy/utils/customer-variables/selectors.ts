import { CustomerVariables, GetVariableResult } from './customer-variables'
import { CustomerVariableType } from './types'

const extractVariable = (result: GetVariableResult) => result.value

export const getAgentUri = async (variables: CustomerVariables) =>
  `/${await getBehaviorPath(variables)}/${await getAgentDownloadPath(variables)}`

export const getResultUri = async (variables: CustomerVariables) =>
  `/${await getBehaviorPath(variables)}/${await getResultPath(variables)}`

export const getStatusUri = async (variables: CustomerVariables) => `/${await getBehaviorPath(variables)}/status`

export const getAgentDownloadPath = async (variables: CustomerVariables) =>
  variables.getVariable(CustomerVariableType.AgentDownloadPath).then(extractVariable)

export const getBehaviorPath = async (variables: CustomerVariables) =>
  variables.getVariable(CustomerVariableType.BehaviourPath).then(extractVariable)

export const getResultPath = async (variables: CustomerVariables) =>
  variables.getVariable(CustomerVariableType.GetResultPath).then(extractVariable)

export const getPreSharedSecret = async (variables: CustomerVariables) =>
  variables.getVariable(CustomerVariableType.PreSharedSecret).then(extractVariable)

export const getFpCdnUrl = async (variables: CustomerVariables) =>
  variables.getVariable(CustomerVariableType.FpCdnUrl).then(extractVariable)

export const getFpIngressBaseHost = async (variables: CustomerVariables) =>
  variables.getVariable(CustomerVariableType.FpIngressBaseHost).then(extractVariable)
