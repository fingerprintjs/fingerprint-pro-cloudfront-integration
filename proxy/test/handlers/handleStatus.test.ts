import { getInMemoryCustomerVariables } from '../utils/customer-variables/in-memory-customer-variables'
import { CustomerVariableType } from '../../utils/customer-variables/types'
import { getStatusInfo, handleStatus } from '../../handlers/handleStatus'

describe('Handle status', () => {
  it('returns correct status info in html if all variables are set', async () => {
    const { customerVariables } = getInMemoryCustomerVariables()

    const result = await handleStatus(customerVariables)

    expect(result.headers).toEqual({
      'content-type': [{ key: 'Content-Type', value: 'text/html' }],
    })

    expect(result.body).toMatchSnapshot()
  })

  it('returns correct status info in html if some variables are using default values', async () => {
    const { customerVariables, variables } = getInMemoryCustomerVariables()

    variables.fpjs_pre_shared_secret = null
    variables.fpjs_agent_download_path = null
    variables.fpjs_get_result_path = null

    const result = await handleStatus(customerVariables)

    expect(result.headers).toEqual({
      'content-type': [{ key: 'Content-Type', value: 'text/html' }],
    })

    expect(result.body).toMatchSnapshot()
  })

  it('returns correct status info in html if some variables are missing', async () => {
    const { customerVariables, variables } = getInMemoryCustomerVariables()

    variables.fpjs_pre_shared_secret = null

    const result = await handleStatus(customerVariables)

    expect(result.headers).toEqual({
      'content-type': [{ key: 'Content-Type', value: 'text/html' }],
    })

    expect(result.body).toMatchSnapshot()
  })
})

describe('Get status info', () => {
  it('returns correct status info', async () => {
    const { customerVariables } = getInMemoryCustomerVariables()

    const result = await getStatusInfo(customerVariables)

    expect(result).toMatchSnapshot()
  })

  it('returns correct response with empty pre shared secret', async () => {
    const { customerVariables, variables } = getInMemoryCustomerVariables()
    variables[CustomerVariableType.PreSharedSecret] = null

    const result = await getStatusInfo(customerVariables)

    expect(result).toMatchSnapshot()
  })

  it('returns correct response with empty non obfuscated variable', async () => {
    const { customerVariables } = getInMemoryCustomerVariables()

    const result = await getStatusInfo(customerVariables)

    expect(result).toMatchSnapshot()
  })
})
