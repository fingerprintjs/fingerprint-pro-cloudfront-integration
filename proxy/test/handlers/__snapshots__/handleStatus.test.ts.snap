// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get status info returns correct response with empty non obfuscated variable 1`] = `
{
  "envInfo": [
    {
      "envVarName": "fpjs_get_result_path",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "result",
    },
    {
      "envVarName": "fpjs_pre_shared_secret",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "********",
    },
    {
      "envVarName": "fpjs_agent_download_path",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "download",
    },
    {
      "envVarName": "fpjs_cdn_url",
      "isInternal": true,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "fpcdn.io",
    },
    {
      "envVarName": "fpjs_ingress_base_host",
      "isInternal": true,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "api.fpjs.io",
    },
  ],
  "styleNonce": "hardcodedStyleNonce",
  "version": "__lambda_func_version__",
}
`;

exports[`Get status info returns correct response with empty pre shared secret 1`] = `
{
  "envInfo": [
    {
      "envVarName": "fpjs_get_result_path",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "result",
    },
    {
      "envVarName": "fpjs_pre_shared_secret",
      "isInternal": false,
      "isSet": false,
      "resolvedBy": null,
      "value": null,
    },
    {
      "envVarName": "fpjs_agent_download_path",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "download",
    },
    {
      "envVarName": "fpjs_cdn_url",
      "isInternal": true,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "fpcdn.io",
    },
    {
      "envVarName": "fpjs_ingress_base_host",
      "isInternal": true,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "api.fpjs.io",
    },
  ],
  "styleNonce": "hardcodedStyleNonce",
  "version": "__lambda_func_version__",
}
`;

exports[`Get status info returns correct status info 1`] = `
{
  "envInfo": [
    {
      "envVarName": "fpjs_get_result_path",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "result",
    },
    {
      "envVarName": "fpjs_pre_shared_secret",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "********",
    },
    {
      "envVarName": "fpjs_agent_download_path",
      "isInternal": false,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "download",
    },
    {
      "envVarName": "fpjs_cdn_url",
      "isInternal": true,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "fpcdn.io",
    },
    {
      "envVarName": "fpjs_ingress_base_host",
      "isInternal": true,
      "isSet": true,
      "resolvedBy": "test provider",
      "value": "api.fpjs.io",
    },
  ],
  "styleNonce": "hardcodedStyleNonce",
  "version": "__lambda_func_version__",
}
`;

exports[`Handle status returns correct status info in html if all variables are set 1`] = `
"<html lang="en-US">
      <head>
        <title>CloudFront integration status</title>
        <meta charset="utf-8">
        <style nonce='hardcodedStyleNonce'>
          body, .env-info {
            display: flex;
          }
          
          body {
            flex-direction: column;
            align-items: center;
          }
          
          body > * {
            margin-bottom: 1em;
          }
        </style>
      </head>
      <body>
        <h1>CloudFront integration status</h1>
        <div>
          Lambda function version: __lambda_func_version__
        </div>
        
      <div>
        ✅ All environment variables are set
      </div>
    
          <span>
            Please reach out our support via <a href="mailto:support@fingerprint.com">support@fingerprint.com</a> if you have any issues
          </span>
      </body>
    </html>"
`;

exports[`Handle status returns correct status info in html if some variables are missing 1`] = `
"<html lang="en-US">
      <head>
        <title>CloudFront integration status</title>
        <meta charset="utf-8">
        <style nonce='hardcodedStyleNonce'>
          body, .env-info {
            display: flex;
          }
          
          body {
            flex-direction: column;
            align-items: center;
          }
          
          body > * {
            margin-bottom: 1em;
          }
        </style>
      </head>
      <body>
        <h1>CloudFront integration status</h1>
        <div>
          Lambda function version: __lambda_func_version__
        </div>
        
    <div class="env-info">
      
        <div class="env-info-item">
            ⚠️ <strong>fpjs_pre_shared_secret </strong> is not defined
        </div>
    </div>
  
          <span>
            Please reach out our support via <a href="mailto:support@fingerprint.com">support@fingerprint.com</a> if you have any issues
          </span>
      </body>
    </html>"
`;

exports[`Handle status returns correct status info in html if some variables are using default values 1`] = `
"<html lang="en-US">
      <head>
        <title>CloudFront integration status</title>
        <meta charset="utf-8">
        <style nonce='hardcodedStyleNonce'>
          body, .env-info {
            display: flex;
          }
          
          body {
            flex-direction: column;
            align-items: center;
          }
          
          body > * {
            margin-bottom: 1em;
          }
        </style>
      </head>
      <body>
        <h1>CloudFront integration status</h1>
        <div>
          Lambda function version: __lambda_func_version__
        </div>
        
    <div class="env-info">
      
        <div class="env-info-item">
            ⚠️ <strong>fpjs_get_result_path </strong> is not defined and uses default value
        </div>
        <div class="env-info-item">
            ⚠️ <strong>fpjs_pre_shared_secret </strong> is not defined
        </div>
        <div class="env-info-item">
            ⚠️ <strong>fpjs_agent_download_path </strong> is not defined and uses default value
        </div>
    </div>
  
          <span>
            Please reach out our support via <a href="mailto:support@fingerprint.com">support@fingerprint.com</a> if you have any issues
          </span>
      </body>
    </html>"
`;
